### PR TITLE
Fix performance regressions in Strings utility

### PR DESCRIPTION
**Pull request by original-brownbear**
Some of the changes in
https://github.com/elastic/elasticsearch/pull/90672 introduced considerable performance hits for many shards benchmarks because the changed methods are used in hot loops here and there. This reverts some of the changes to the previous code and delegates to the fast `isNullOrBlank` to fix `hasText` performance for strings.

E.g. `hasText` which hardly registered in profiling before the above PR went to 2%+ of total CPU usage during snapshotting:
image

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/Strings.java [hunk 1]
```diff
@@ -137,7 +137,13 @@ public static boolean hasText(CharSequence str) {
         if (hasLength(str) == false) {
             return false;
         }
-        return str.chars().anyMatch(c -> Character.isWhitespace(c) == false);
+        int strLen = str.length();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/Strings.java:140]: This is probably not hot much after this change, but I still find it weird to have a utility like this that is "slow" like this.

File: server/src/main/java/org/elasticsearch/common/Strings.java [hunk 2]
```diff
@@ -286,11 +292,22 @@ private static String changeFirstCharacterCase(String str, boolean capitalize) {
         .collect(Collectors.joining(",", "[", "]"));
 
     public static boolean validFileName(String fileName) {
-        return fileName.chars().noneMatch(c -> isInvalidFileNameCharacter((char) c));
+        for (int i = 0; i < fileName.length(); i++) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/Strings.java:295]: This is hot in a couple of spots, the `validFileNameExcludingAstrix` isn't but I figured if I fix one I might as well fix both.

**Comment by thecoop** [comment]
Could we have some comments indicating these are hot methods & so need to be left as-is?

**Comment by thecoop** [comment]
The fact of streams being so slow makes me very sad :(

**Comment by original-brownbear** [comment]
> Could we have some comments indicating these are hot methods & so need to be left as-is?

idk, to me it seems like we should just not regress the performance of utility methods like that and build them fast by default? It seems like a lot of clutter to start requesting people not slow down methods on utility methods?

**Comment by thecoop** [comment]
> > Could we have some comments indicating these are hot methods & so need to be left as-is?
> 
> idk, to me it seems like we should just not regress the performance of utility methods like that and build them fast by default? It seems like a lot of clutter to start requesting people not slow down methods on utility methods?

It might be helpful to the next person who comes along and looks to simplify these methods, to note that they are used in hot paths and so should be left alone, to save them the effort of trying it & noting it drops performance.

**Review by thecoop — APPROVED** [review]

**Comment by original-brownbear** [comment]
thanks Simon!

File: server/src/main/java/org/elasticsearch/common/Strings.java [hunk 3]
```diff
@@ -137,7 +137,13 @@ public static boolean hasText(CharSequence str) {
...
             return false;
         }
-        return str.chars().anyMatch(c -> Character.isWhitespace(c) == false);
+        int strLen = str.length();
+        for (int i = 0; i < strLen; i++) {
+            if (Character.isWhitespace(str.charAt(i)) == false) {
+                return true;
+            }
+        }
+        return false;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/Strings.java:146]: I wonder why don't we use `java.lang.String#isBlank` instead here?

**Pull request merged**